query-algebrizer/src/clauses/where_fn.rs

```diff
@@ -0,0 +1,384 @@
+    fn into_typed_value(self, lookup: F) -> Result<TypedValue> {
+        match self {
+            FnArg::EntidOrInteger(x) => Ok(TypedValue::Long(x)),
+            FnArg::Ident(x) => Ok(TypedValue::Keyword(Rc::new(x))),
+            FnArg::Constant(NonIntegerConstant::Boolean(x)) => Ok(TypedValue::Boolean(x)),
+            FnArg::Constant(NonIntegerConstant::BigInteger(_)) => unimplemented!(),
+            FnArg::Constant(NonIntegerConstant::Float(x)) => Ok(TypedValue::Double(x)),
+            FnArg::Constant(NonIntegerConstant::Text(x)) => Ok(TypedValue::String(x)),
+            FnArg::Variable(var) => lookup(var),
+            FnArg::Vector(_) |
+            FnArg::SrcVar(_) => bail!(ErrorKind::InvalidGroundConstant),
+        }
+    }
+
+    fn into_column_vector(self, lookup: F) -> Result<Vec<TypedValue>> {
+        if let FnArg::Vector(children) = self {
+            children.into_iter().map(|x| x.into_typed_value(&lookup)).collect()
+        } else {
+            bail!(ErrorKind::InvalidGroundConstant)
+        }
+    }
+
+    /// Convert an atom into a 1x1 matrix.
+    fn into_unit_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>> {
+        self.into_typed_value(lookup).map(|x| vec![vec![x]])
+    }
+
+    /// Convert a vector of atoms into a 1xN matrix.
+    fn into_row_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>> {
+        self.into_column_vector(lookup).map(|x| vec![x])
+    }
+
+    /// Convert a vector of atoms into an Mx1 matrix.
+    fn into_column_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>> {
+        self.into_column_vector(lookup).map(|x| x.into_iter().map(|y| vec![y]).collect())
+    }
+
+    /// Convert a vector of vectors (each of atoms) into an MxN matrix.
+    fn into_block_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>> {
+        if let FnArg::Vector(children) = self {
+            children.into_iter().map(|x| x.into_column_vector(&lookup)).collect()
+        } else {
+            bail!(ErrorKind::InvalidGroundConstant)
+        }
+    }
+}
+
+/// This function takes a vector of rows of possibly different lengths and a sequence of columns to
+/// ignore.  It ensures that each row has the same length, and drops the columns that have `false`
+/// in the `column_is_bound` sequence.  Finally, it ensures that each column is homogeneously typed,
+/// and returns a sequence of the derived type for each column.
+fn reduced_rows_and_column_types(rows: Vec<Vec<TypedValue>>, column_is_bound: &Vec<bool>) -> Result<(Vec<Vec<TypedValue>>, Vec<ValueType>)> {
+    let mut matrix = Vec::with_capacity(rows.len());
+
+    for row in rows {
+        if column_is_bound.len() != row.len() {
+            bail!(ErrorKind::InvalidGroundConstant)
+        }
+
+        let mut retained_row = Vec::with_capacity(column_is_bound.len());
+        for (is_bound, value) in column_is_bound.into_iter().zip(row.into_iter()) {
+            if *is_bound {
+                retained_row.push(value);
```

Comment:
Given that `rows` is passed owned, I would rather see this implemented as in-place modification:

```
let expected = column_is_bound.len();
let to_remove = column_is_bound.iter()
                               .enumerate()
                               .rev()         // So we work from the end for remove.
                               .filter_map(|i, is_bound| if is_bound { Some(i) } else { None })
                               .collect();
for row in rows {
    if expected != row.len() {
        bail!(…);
    }
    for i in to_remove {
        row.remove(*i);
    }
    matrix.push(row);
}
```